wtxmgr/query.go

```diff
@@ -0,0 +1,462 @@
+		v := existsRawUnmined(ns, txHash[:])
+		if v != nil {
+			details, err = s.unminedTxDetails(ns, txHash, v)
+			return err
+		}
+
+		// Otherwise, if there exists a mined transaction with this matching
+		// hash, skip over to the newest and begin fetching all details.
+		k, v := latestTxRecord(ns, txHash)
+		if v == nil {
+			// not found
+			return nil
+		}
+
+		details, err = s.txDetails(ns, txHash, k, v)
+		return err
+	})
+	return details, err
+}
+
+// UniqueTxDetails looks up all recorded details for a transaction recorded
+// mined in some particular block, or an unmined transaction if block is nil.
+//
+// Not finding a transaction with this hash from this block is not an error.  In
+// this case, a nil TxDetails is returned.
+func (s *Store) UniqueTxDetails(txHash *wire.ShaHash, block *Block) (*TxDetails, error) {
+	var details *TxDetails
+	err := scopedView(s.namespace, func(ns walletdb.Bucket) error {
+		var err error
+		if block == nil {
+			v := existsRawUnmined(ns, txHash[:])
+			if v == nil {
+				return nil
+			}
+			details, err = s.unminedTxDetails(ns, txHash, v)
+			return err
+		}
+
+		k, v := existsTxRecord(ns, txHash, block)
+		if v == nil {
+			return nil
+		}
+		details, err = s.txDetails(ns, txHash, k, v)
+		return err
+	})
+	return details, err
+}
+
+func (s *Store) txDetails(ns walletdb.Bucket, txHash *wire.ShaHash, recKey, recVal []byte) (*TxDetails, error) {
+	var details TxDetails
+
+	// Parse transaction record k/v, lookup the full block record for the
+	// block time, and read all matching credits, debits.
+	err := readRawTxRecord(txHash, recVal, &details.TxRecord)
+	if err != nil {
+		return nil, err
+	}
+	err = readRawTxRecordBlock(recKey, &details.Block.Block)
+	if err != nil {
+		return nil, err
+	}
+	details.Block.Time, err = fetchBlockTime(ns, details.Block.Height)
+	if err != nil {
+		return nil, err
+	}
+
+	credIter := makeCreditIterator(ns, recKey)
+	for credIter.next() {
+		if int(credIter.elem.Index) >= len(details.MsgTx.TxOut) {
+			str := "saved credit index exceeds number of outputs"
+			return nil, storeError(ErrData, str, nil)
+		}
+
+		// The credit iterator does not record whether this credit was
+		// spent by an unmined transaction, so check that here.
+		if !credIter.elem.Spent {
+			k := canonicalOutPoint(txHash, credIter.elem.Index)
+			spent := existsRawUnminedInput(ns, k) != nil
+			credIter.elem.Spent = spent
+		}
+		details.Credits = append(details.Credits, credIter.elem)
+	}
+	if credIter.err != nil {
+		return nil, credIter.err
+	}
+
+	debIter := makeDebitIterator(ns, recKey)
+	for debIter.next() {
+		if int(debIter.elem.Index) >= len(details.MsgTx.TxIn) {
+			str := "saved debit index exceeds number of inputs"
+			return nil, storeError(ErrData, str, nil)
+		}
+
+		details.Debits = append(details.Debits, debIter.elem)
+	}
+	return &details, debIter.err
+}
+
+func (s *Store) unminedTxDetails(ns walletdb.Bucket, txHash *wire.ShaHash, v []byte) (*TxDetails, error) {
```

Comment:
This would also be better defined before it's called by `TxDetails` and `UniqueTxDetails`.

Also, the code in `TxDetails` and `UniqueTxDetails` invokes this function before `txDetails`, so it should ideally be defined before `txDetails` as well.